src/main/java/io/vertx/core/net/impl/pool/NonBlockingSynchronization2.java

```diff
@@ -0,0 +1,43 @@
+package io.vertx.core.net.impl.pool;
+
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class NonBlockingSynchronization2<S> implements Synchronization<S> {
+
+  private final Queue<Action<S>> q = PlatformDependent.newMpscQueue(Integer.MAX_VALUE);
+  private final AtomicInteger s = new AtomicInteger();
+  private final S state;
+
+  public NonBlockingSynchronization2(S state) {
+    this.state = state;
+  }
+
+  @Override
+  public void execute(Action<S> action) {
+    q.add(action);
+    while (true) {
+      int v = s.get();
+      if (v == 0) {
+        if (s.compareAndSet(0, 1)) {
```

Comment:
I see an unecessary op here: it's ok just for the very first attempt of `s.compareAndSet(0, 1)` that we fail fast and just move on, terminating `execute`, while is mandatory for the other subsequent loop iteration that we check for `q::isEmpty`. 
I would group 
```java
          Action<S> a;
          while ((a = q.poll()) != null) {
            Runnable post = a.execute(state);
            if (post != null) {
              post.run();
            }
          }
```
into its own method eg `pollAndExecute` and will separate the 2 cases:
```java
q.add
if (s.get() != 0 || !s.compareAndSet(0, 1)) {     
    return;
}
do {
    try {
        pollAndExecute(action);
    } finally {
        s.set(0);
    }
} while (!q.isEmpty() && s.compareAndSet(0,1));
```

Reply:
Just thinking loud and not related with this PR, but if we pack the loop logic:
```java
do {
    try {
        pollAndExecute(q, action);
    } finally {
        s.set(0);
    }
} while (!q.isEmpty() && s.compareAndSet(0, 1));
```
into a separate method ie `lockedPollAndExecute` and we get an event loop group instance, we can just send back a continuation (with s.get() == 1, to save other threads to try the same) instead of keeping a single event loop thread busy for "too long" (and I would add a check with a configurable busy time eg 100 us).